### PR TITLE
add MQTT preset support for Climate components

### PR DESCRIPTION
**Pull request by jmichiel**
# What does this implement/fix?

The current preset state and changing presets through MQTT was still lacking.
ClimateTraits::get_supports_presets() was also updated to be in line with ClimateTraits::get_supports_fan_modes() behaviour.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation:** esphome/esphome-docs#2641

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

Tested with a dummy custom climate: discovery is correct, state changes are published and commands forwarded.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/mqtt/mqtt_climate.cpp [hunk 1]
```diff
@@ -66,12 +66,36 @@ void MQTTClimateComponent::send_discovery(JsonObject root, mqtt::SendDiscoveryCo
   // temperature units are always coerced to Celsius internally
   root[MQTT_TEMPERATURE_UNIT] = "C";
 
-  if (traits.supports_preset(CLIMATE_PRESET_AWAY)) {
-    // away_mode_command_topic
-    root[MQTT_AWAY_MODE_COMMAND_TOPIC] = this->get_away_command_topic();
-    // away_mode_state_topic
-    root[MQTT_AWAY_MODE_STATE_TOPIC] = this->get_away_state_topic();
+  if (traits.get_supports_presets() || !traits.get_supported_custom_presets().empty()) {
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_climate.cpp:69]: You added this check inside `get_supports_presets()`
```suggestion
  if (traits.get_supports_presets()) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/climate/climate_traits.h [hunk 2]
```diff
@@ -107,7 +107,7 @@ class ClimateTraits {
   void add_supported_preset(ClimatePreset preset) { supported_presets_.insert(preset); }
   void add_supported_custom_preset(const std::string &preset) { supported_custom_presets_.insert(preset); }
   bool supports_preset(ClimatePreset preset) const { return supported_presets_.count(preset); }
-  bool get_supports_presets() const { return !supported_presets_.empty(); }
+  bool get_supports_presets() const { return !supported_presets_.empty() || !supported_custom_presets_.empty(); }
```
Comment (jesserockz) [inline esphome/components/climate/climate_traits.h]: Just had a quick browse through the code and this change might break existing usages where it is also checking for custom in a different conditional

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/climate/climate_traits.h (hunk 2 above)
Comment (jmichiel) [reply esphome/components/climate/climate_traits.h]: I feared as much and perhaps that's why it wasn't aligned with get_supports_fan_modes?

Comment (jmichiel) [reply esphome/components/climate/climate_traits.h]: I reverted the changes

**Review by jesserockz — APPROVED** [review]

**Pull request merged**